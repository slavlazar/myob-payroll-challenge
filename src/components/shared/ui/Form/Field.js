import React, { memo } from 'react';
import PropTypes from 'prop-types';

import {
  Group,
  LabelGroup,
  Label,
  InputGroup,
  ErrorMessage,
} from './Field.style';

/**
 * Renders a form field with label (if present), along
 * with Component (if present), or render prop function -
 * also handles error messaging if field has an error returned
 * from Formik schema validation
 * @param {object} Object of input props and render function or component
 */
const InputComponent = ({
  component: Component,
  render,
  id,
  name,
  label,
  errorMessage,
  ...inputProps
}) => {
  const htmlId = id || name;

  const errorAriaDescribe = `${htmlId}_error`;

  return (
    <Group>
      {label && (
        <LabelGroup>
          <Label className="form-label" htmlFor={htmlId}>
            {label}
          </Label>
        </LabelGroup>
      )}
      <InputGroup>
        {Component ? (
          <Component
            id={htmlId}
            name={name}
            error={errorMessage}
            aria-describedby={errorAriaDescribe}
            {...inputProps}
          />
        ) : (
          render({
            id: htmlId,
            name,
            errorMessage,
            'aria-describedby': errorAriaDescribe,
            ...inputProps,
          })
        )}
        {errorMessage && (
          <ErrorMessage className="form-error-message" id={errorAriaDescribe}>
            {errorMessage}
          </ErrorMessage>
        )}
      </InputGroup>
    </Group>
  );
};

InputComponent.propTypes = {
  component: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
  render: PropTypes.func,
  id: PropTypes.string,
  name: PropTypes.string,
  label: PropTypes.string,
  errorMessage: PropTypes.string,
};

InputComponent.defaultProps = {
  component: null,
  render: () => {},
  id: null,
  name: null,
  label: null,
  errorMessage: null,
};

// Prevent re-render if value or errorMessage are the same
export default memo(
  InputComponent,
  (prevProps, nextProps) =>
    prevProps.value === nextProps.value &&
    prevProps.errorMessage === nextProps.errorMessage
);
