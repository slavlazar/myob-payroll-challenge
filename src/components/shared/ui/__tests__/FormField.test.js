import React from 'react';
import { shallow } from 'enzyme';

import Field from '../Form/Field';

describe('Components: Shared: UI: Form Field', () => {
  it('renders correctly', () => {
    const wrapped = shallow(<Field />);
    expect(wrapped).toMatchSnapshot();
  });

  it('should render a label', () => {
    const label = 'Test label';
    const wrapped = shallow(<Field label={label} />);
    expect(wrapped.find('.form-label').text()).toEqual(label);
  });

  it('should display an error message', () => {
    const errorMessage = 'Error Message!';
    const wrapped = shallow(<Field errorMessage={errorMessage} />);
    expect(wrapped.find('.form-error-message').text()).toEqual(errorMessage);
  });

  it('should render passed in component', () => {
    const component = () => <div>Testing</div>;
    const wrapped = shallow(<Field component={component} />);
    expect(wrapped.find(component)).toHaveLength(1);
  });
});
