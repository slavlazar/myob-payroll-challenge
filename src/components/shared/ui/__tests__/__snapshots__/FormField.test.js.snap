// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components: Shared: UI: Form Field renders correctly 1`] = `
<styled.div>
  <styled.div />
</styled.div>
`;
